File: app/rust/accounting-service/financial_accounting/src/domain/balance/monthly_account_balance.rs
use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

/// 月次勘定科目残高エンティティ
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MonthlyAccountBalance {
    /// 決算期（会計年度）
    pub fiscal_year: i32,

    /// 月度（1〜12）
    pub month: i32,

    /// 勘定科目コード
    pub account_code: String,

    /// 補助科目コード
    pub sub_account_code: String,

    /// 部門コード
    pub department_code: String,

    /// プロジェクトコード
    pub project_code: String,

    /// 決算仕訳フラグ
    pub settlement_flag: i32,

    /// 月初残高
    pub beginning_balance: Decimal,

    /// 借方金額
    pub debit_amount: Decimal,

    /// 貸方金額
    pub credit_amount: Decimal,

    /// 月末残高
    pub ending_balance: Decimal,

    /// 作成日時
    pub created_at: DateTime<Utc>,

    /// 更新日時
    pub updated_at: DateTime<Utc>,
}

impl MonthlyAccountBalance {
    /// 月末残高を計算（月初残高 + 借方金額 - 貸方金額）
    pub fn calculate_ending_balance(&self) -> Decimal {
        self.beginning_balance + self.debit_amount - self.credit_amount
    }

    /// 当月の純増減額（借方金額 - 貸方金額）
    pub fn get_net_change(&self) -> Decimal {
        self.debit_amount - self.credit_amount
    }

    /// 複合主キーを抽出
    pub fn key(&self) -> MonthlyAccountBalanceKey {
        MonthlyAccountBalanceKey {
            fiscal_year: self.fiscal_year,
            month: self.month,
            account_code: self.account_code.clone(),
            sub_account_code: self.sub_account_code.clone(),
            department_code: self.department_code.clone(),
            project_code: self.project_code.clone(),
            settlement_flag: self.settlement_flag,
        }
    }
}

/// 月次勘定科目残高の複合主キー
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct MonthlyAccountBalanceKey {
    pub fiscal_year: i32,
    pub month: i32,
    pub account_code: String,
    pub sub_account_code: String,
    pub department_code: String,
    pub project_code: String,
    pub settlement_flag: i32,
}

File: app/ruby/accounting_system/app/models/monthly_account_balance.rb
# frozen_string_literal: true

# 月次勘定科目残高モデル
# 月ごとの月初残高・借方金額・貸方金額・月末残高を記録する
class MonthlyAccountBalance < ApplicationRecord
  # アソシエーション
  belongs_to :account, foreign_key: :account_code, primary_key: :code

  # バリデーション
  validates :fiscal_year, presence: true, numericality: { only_integer: true, greater_than: 0 }
  validates :month, presence: true, inclusion: { in: 1..12 }
  validates :account_code, presence: true, length: { maximum: 10 }
  validates :sub_account_code, length: { maximum: 10 }
  validates :department_code, length: { maximum: 5 }
  validates :project_code, length: { maximum: 10 }
  validates :settlement_flag, inclusion: { in: [0, 1] }
  validates :beginning_balance, numericality: true
  validates :debit_amount, numericality: { greater_than_or_equal_to: 0 }
  validates :credit_amount, numericality: { greater_than_or_equal_to: 0 }
  validates :ending_balance, numericality: true

  # スコープ
  scope :normal_entries, -> { where(settlement_flag: 0) }
  scope :settlement_entries, -> { where(settlement_flag: 1) }
  scope :by_fiscal_year, ->(year) { where(fiscal_year: year) }
  scope :by_month, ->(month) { where(month: month) }
  scope :by_account, ->(account_code) { where(account_code: account_code) }
  scope :by_department, ->(department_code) { where(department_code: department_code) }
  scope :by_project, ->(project_code) { where(project_code: project_code) }

  # 月末残高を計算（月初残高 + 借方金額 - 貸方金額）
  def calculate_ending_balance
    beginning_balance + debit_amount - credit_amount
  end
end
